clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java

```diff
@@ -268,6 +287,65 @@ private Result anotherNodeInAnotherGroupHasWantedState(StorageNodeInfo nodeInfo)
+     */
+    private Optional<Result> otherNodesHaveWantedState(StorageNodeInfo nodeInfo, String newDescription, ClusterState clusterState) {
+        Node node = nodeInfo.getNode();
+
+        if (groupVisiting.isHierarchical()) {
+            Set<Integer> groupsWithNodesWantedStateNotUp = groupsWithUserWantedStateNotUp();
+            if (groupsWithNodesWantedStateNotUp.size() == 0) {
+                log.log(FINE, "groupsWithNodesWantedStateNotUp=0");
+                return Optional.empty();
+            }
+
+            Set<Integer> groupsWithSameStateAndDescription = groupsWithSameStateAndDescription(MAINTENANCE, newDescription);
+            if (aGroupContainsNode(groupsWithSameStateAndDescription, node)) {
+                log.log(FINE, "Node is in group with same state and description, allow");
+                return Optional.of(allowSettingOfWantedState());
+            }
+            // There are groups with nodes not up, but with another description, probably operator set
+            if (groupsWithSameStateAndDescription.size() == 0) {
+                return Optional.of(createDisallowed("Wanted state already set for another node in groups: " +
+                                                            sortSetIntoList(groupsWithNodesWantedStateNotUp)));
+            }
+
+            Set<Integer> retiredAndNotUpGroups = groupsWithNotRetiredAndNotUp(clusterState);
+            int numberOfGroupsToConsider = retiredAndNotUpGroups.size();
+            // Subtract one group if node is in a group with nodes already retired or not up, since number of such groups wil
+            // not increase if we allow node to go down
+            if (aGroupContainsNode(retiredAndNotUpGroups, node)) {
+                numberOfGroupsToConsider = retiredAndNotUpGroups.size() - 1;
+            }
+            if (numberOfGroupsToConsider < maxNumberOfGroupsAllowedToBeDown) {
+                log.log(FINE, "Allow, retiredAndNotUpGroups=" + retiredAndNotUpGroups);
+                return Optional.of(allowSettingOfWantedState());
+            }
+
+            return Optional.of(createDisallowed(String.format("At most %d groups can have wanted state: %s",
+                                                              maxNumberOfGroupsAllowedToBeDown,
+                                                              sortSetIntoList(retiredAndNotUpGroups))));
+        } else {
+            // Return a disallow-result if there is another node with a wanted state
+            var otherNodeHasWantedState = otherNodeHasWantedState(nodeInfo);
+            if ( ! otherNodeHasWantedState.settingWantedStateIsAllowed())
+                return Optional.of(otherNodeHasWantedState);
+        }
+        return Optional.empty();
+    }
+
+    private ArrayList<Integer> sortSetIntoList(Set<Integer> retiredOrNotUpGroups) {
+        var groupsWithNodesDown = new ArrayList<>(retiredOrNotUpGroups);
```

Comment:
Should the parameter and variable names be unified here? It could be a bit confusing how `retiredOrNotUpGroups` suddenly becomes `groupsWithNodesDown`.

Reply:
Yes, fixed